services/user/user.go

```diff
@@ -27,19 +28,94 @@ import (
 // DeleteUser completely and permanently deletes everything of a user,
 // but issues/comments/pulls will be kept and shown as someone has been deleted,
 // unless the user is younger than USER_DELETE_WITH_COMMENTS_MAX_DAYS.
-func DeleteUser(u *user_model.User) error {
+func DeleteUser(ctx context.Context, u *user_model.User, purge bool) error {
 	if u.IsOrganization() {
 		return fmt.Errorf("%s is an organization not a user", u.Name)
 	}
 
+	if purge {
+		// Disable the user first
+		if err := user_model.UpdateUserCols(ctx, &user_model.User{
```

Comment:
`ctx` should contain a transaction so that it could fall back or we need a function to update user roll back.

Reply:
This needs to apply immediately to prevent any further login or action by the user being deleted. They need to be restricted and prohibited from further login.

I can add a rollback function that would unlock them but the point of this is ensure that the user is ready to be purged.

Reply:
Then perhaps the comment should be adapted to indicate exactly this reason.

Reply:
Actually I don't think there should be a rollback. The user was intended to be purged - they should remain locked out.